gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java

```diff
@@ -191,59 +197,162 @@ private String dependencyPathsOfProblematicJars(
+      List<String> dependencyPathElementsReversed = new ArrayList<>();
+      for (ResolvedComponentResultNode iter = this; iter != null; iter = iter.parent) {
+        dependencyPathElementsReversed.add(formatComponentResult(iter.componentResult));
+      }
+      Collections.reverse(dependencyPathElementsReversed);
+      String dependencyPath = Joiner.on(" / ").join(dependencyPathElementsReversed);
+      return dependencyPath;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("componentResult", componentResult)
+          .add("parent", parent)
+          .toString();
     }
 
-    for (DependencyResult dependencyResult : item.getDependencies()) {
-      if (dependencyResult instanceof ResolvedDependencyResult) {
-        ResolvedDependencyResult resolvedDependencyResult =
-            (ResolvedDependencyResult) dependencyResult;
-        ResolvedComponentResult child = resolvedDependencyResult.getSelected();
+    @Override
+    public boolean equals(Object other) {
+      if (this == other) {
+        return true;
+      }
+      if (other == null || getClass() != other.getClass()) {
+        return false;
+      }
+      ResolvedComponentResultNode that = (ResolvedComponentResultNode) other;
+      return Objects.equals(componentResult, that.componentResult)
+          && Objects.equals(parent, that.parent);
+    }
 
-        if (stack.contains(child)) {
-          // Circular dependency check
-          if (checkedCircularDependency.add(child)) {
+    @Override
+    public int hashCode() {
+      return Objects.hash(componentResult, parent);
+    }
+  }
+
+  /**
+   * Returns mapping from Maven coordinates to their dependency paths appearing in the dependency
+   * graph
+   *
+   * @param componentResult The root project in the dependency graph
+   * @param targetCoordinatesSet The Maven coordinates to check their dependency paths
+   */
+  private ListMultimap<String, String> groupCoordinatesToDependencyPaths(
+      ResolvedComponentResult componentResult, Set<String> targetCoordinatesSet) {
+
+    ListMultimap<String, String> coordinatesToDependencyPaths =
+        MultimapBuilder.hashKeys().arrayListValues().build();
+    // No need to print the same circular dependency multiple times
+    Set<ResolvedComponentResult> checkedCircularDependency = new HashSet<>();
+
+    for (String targetCoordinates : targetCoordinatesSet) {
+      // Queue of dependnecy nodes. Each node knows its parent.
+      ArrayDeque<ResolvedComponentResultNode> queue = new ArrayDeque<>();
+      ResolvedComponentResultNode firstItem =
+          new ResolvedComponentResultNode(componentResult, null);
+      queue.add(firstItem);
+
+      // Mapping to omit duplicate dependency paths in the output. This is a mapping from Maven
+      // coordinates to dependency nodes that has the dependency of the coordinate in their direct
+      // or transitive dependencies.
```

Comment:
This comment was outdated. Updated as

```
      // A set to omit duplicate dependency paths in the output. When a node is found to be in this
      // set while traversing the graph, we do not need to check the children, because we know that
      // the dependency paths from that node to the targetCoordinates are already added to
      // coordinatesToDependencyPaths.
      Set<ResolvedComponentResult> nodesDependOnTarget = new HashSet<>();
```

Reply:
I didn't think of a such `findDependencies()` that would improve code here. But newly added `getDependencies` should give more clarity on the code.